File: web/src/api/tiers.ts
// Tiers (docs/design/phase2-3.md §8, §13): the ordered rule set and its revision, the presets and
// condition fields of the rule editor, the preview ("what gets backed up and why"), the manual
// flags, the Library item view (GET /catalog/files/{id}) and the release of kept files.
//
// D1 (the user's decision): with no rules every file is full. A preset is only loaded into the
// editor; nothing here saves rules except saveTierRules, which the user triggers.

import { api, query, toPaged } from './client';
import type { Job, Paged } from './types';

/** Tier is what a destination does with a file: copy it, only list it in manifests, or neither. */
export type Tier = 'full' | 'manifest' | 'skip';

export const TIERS: Tier[] = ['full', 'manifest', 'skip'];

/** TierMatch says how a rule's conditions combine. */
export type TierMatch = 'all' | 'any';

/** TierCondition is one test of a rule; value is absent for operators that take none (never). */
export interface TierCondition {
  field: string;
  op: string;
  value?: unknown;
}

/** TierRule is a stored rule (GET /tiers/rules). */
export interface TierRule {
  id: number;
  priority: number;
  name: string;
  enabled: boolean;
  match: TierMatch;
  conditions: TierCondition[];
  action: Tier;
  /** "Applies at": null is every destination, a list only those, [] none. */
  destinationIds: number[] | null;
  createdAt: string;
  updatedAt: string;
}

/** TierRuleInput is a rule as PUT /tiers/rules, the preview draft and the presets carry it. */
export interface TierRuleInput {
  /** Absent (or 0) for a new rule. */
  id?: number;
  name: string;
  enabled?: boolean;
  match?: TierMatch;
  conditions: TierCondition[];
  action: Tier;
  destinationIds: number[] | null;
}

export interface TierRuleSet {
  revision: number;
  rules: TierRule[];
}

/** TierWarning is a rule value no fresh index knows (a renamed tag or profile, §8.7). Indexes are 0-based. */
export interface TierWarning {
  ruleIndex: number;
  conditionIndex: number;
  message: string;
}

export interface TierRulesSaved extends TierRuleSet {
  warnings: TierWarning[];
}

export interface TierPreset {
  id: string;
  name: string;
  description: string;
  rules: TierRuleInput[];
}

export type TierValueType = 'bool' | 'string' | 'int' | 'ints';

export interface TierSuggestion {
  value: unknown;
  label: string;
}

/** TierField is a condition field of the editor (GET /tiers/fields). */
export interface TierField {
  field: string;
  label: string;
  ops: string[];
  valueType: TierValueType;
  /** Operators that take no value (tautulli.lastWatched never). */
  noValueOps?: string[];
  /** bytes, days or plays for an int value. */
  unit?: string;
  /** catalog, arr, plex, tautulli, seerr, maintainerr or flag. */
  source: string;
  available: boolean;
  /** Why the field is not available (D16: its provider is not built yet). */
  reason?: string;
  suggestions: TierSuggestion[];
}

export type TierResult = 'true' | 'false' | 'unknown';

/** TierReason is one evaluated condition (§8.1). A Seerr user appears by id only. */
export interface TierReason {
  ruleId: number;
  /** -1 for the built-in irreplaceable override. */
  conditionIndex: number;
  field: string;
  op: string;
  value?: unknown;
  /** The fact compared: a tag list, a profile, a size, a date, plays. */
  actual?: unknown;
  result: TierResult;
  source: { kind: string; integrationId?: number };
  /** Why the result is unknown ("Radarr cache is 31 h old"). */
  why?: string;
}

/** TierCount is a number of files and their bytes. */
export interface TierCount {
  files: number;
  bytes: number;
}

export interface TierFullCount extends TierCount {
  /** A hardlink group's content counted once. */
  uniqueBytes: number;
}

export interface TierRuleCount {
  /** 0 for the built-ins; a draft's new rules have the negative of their position (-1 first). */
  ruleId: number;
  name: string;
  action: Tier;
  files: number;
  bytes: number;
}

export interface TierConfigBackup {
  kind: 'plexdb' | 'arr';
  integrationId: number;
  name: string;
  lastBytes: number;
}

export interface TierDestinationPreview {
  destinationId: number;
  name: string;
  stored: TierCount;
  full: TierFullCount;
  manifest: TierCount;
  skip: TierCount;
  unknownPromoted: TierCount;
  toCopy: TierCount;
  kept: TierCount;
  movedToNonFull: TierCount;
  byRule: TierRuleCount[] | null;
  configBackups: TierConfigBackup[] | null;
}

export interface TierUnknownSource {
  integrationId: number;
  name: string;
  reason: string;
}

/** TierPreview is POST /tiers/preview: the saved rules (revision) or a draft ("draft"). */
export interface TierPreview {
  id: string;
  revision: number | 'draft';
  createdAt: string;
  unknownSources: TierUnknownSource[] | null;
  staleReferences: TierWarning[] | null;
  destinations: TierDestinationPreview[] | null;
}

export type TierItemState = 'stored' | 'to-copy' | 'kept' | 'not-copied';

export interface TierPreviewItem {
  destinationId: number;
  fileId: number;
  sourceId: number;
  sourceName: string;
  relPath: string;
  size: number;
  tier: Tier;
  ruleId: number;
  ruleName: string;
  reasons: TierReason[] | null;
  unknown: TierReason[] | null;
  unknownPromoted: boolean;
  /** The file whose decision this is (a sidecar's media file, a hardlink group's name). */
  follows?: string;
  state: TierItemState;
}

export interface TierPreviewItemQuery {
  destinationId?: number;
  tier?: Tier | '';
  /** The deciding rule (0: the built-ins); undefined for every rule. */
  ruleId?: number;
  state?: TierItemState | '';
  search?: string;
  page: number;
  pageSize: number;
}

export interface ExternalIds {
  tmdb?: number;
  imdb?: string;
  tvdb?: number;
  tvmaze?: number;
  mbid?: string;
}

/** ItemFlag is a manual flag (GET /tiers/flags): an *arr item or a path. */
export interface ItemFlag {
  id: number;
  flag: 'irreplaceable';
  kind: 'arr' | 'path';
  integrationId: number | null;
  arrKind?: string;
  arrId?: number;
  externalIds: ExternalIds;
  lastSourceId: number | null;
  lastRelPath: string | null;
  sourceId: number | null;
  /** '' is the whole source. */
  relPath: string | null;
  note: string;
  createdAt: string;
  updatedAt: string;
  resolved: boolean;
  reason?: string;
}

export type FlagTarget = { integrationId: number; kind: string; arrId: number } | { sourceId: number; relPath: string };

export interface FlagInput {
  flag: 'irreplaceable';
  target: FlagTarget;
  note: string;
}

// ---- The Library item view (GET /catalog/files/{id}) ----

export interface ArrItemFacts {
  integrationId: number;
  app: string;
  itemId: number;
  /** movie, series or artist. */
  kind: string;
  arrId: number;
  title: string;
  year: number;
  externalIds: ExternalIds;
  tags: string[] | null;
  qualityProfile: string;
  qualityProfileId: number;
  rootFolder: string;
  monitored: boolean;
  genres: string[] | null;
  folder?: string;
}

export interface ArrFacts {
  state: 'item' | 'unmanaged' | 'unknown';
  why?: string;
  integrationId?: number;
  item?: ArrItemFacts;
  /** Attributed by folder (an extra): episode-level facts are unknown. */
  byFolder?: boolean;
  arrFileId?: number;
  dateAdded?: string;
  quality?: string;
}

export interface PlexFacts {
  known: boolean;
  why?: string;
  integrationId?: number;
  section?: string;
  addedAt?: string;
}

export interface WatchFacts {
  known: boolean;
  why?: string;
  integrationId?: number;
  plays: number;
  lastWatched?: string;
  lowerBound?: boolean;
}

export interface RequestFacts {
  requested: TierResult;
  why?: string;
  integrationId?: number;
  users: number[] | null;
}

export interface MaintainerrFacts {
  pending: TierResult;
  why?: string;
  integrationId?: number;
  deleteAfter?: string;
}

export interface FileFacts {
  arr: ArrFacts;
  /** null: nothing supplies these facts (their conditions are unknown). */
  plex: PlexFacts | null;
  watch: WatchFacts | null;
  requests: RequestFacts | null;
  maintainerr: MaintainerrFacts | null;
  /** The irreplaceable flags that cover the file. */
  flags: number[] | null;
  /** A sidecar's media file, whose decision it takes. */
  follows?: string;
  unknown: { source: string; reason: string }[] | null;
}

export interface FileRecord {
  state: string;
  relPath: string;
  size: number;
  copiedAt: string | null;
  verifiedAt: string | null;
}

export interface FileTierAtDestination {
  destinationId: number;
  destinationName: string;
  tier: Tier;
  ruleId: number;
  ruleName: string;
  reasons: TierReason[] | null;
  unknown: TierReason[] | null;
  unknownPromoted: boolean;
  follows?: string;
  record: FileRecord | null;
}

export interface FileDetail {
  file: { id: number; relPath: string; size: number; mtime: string; hardlinkGroup: string | null };
  source: { id: number; name: string };
  facts: FileFacts;
  tiers: FileTierAtDestination[] | null;
}

// ---- Calls ----

export function getTierRules(): Promise<TierRuleSet> {
  return api<TierRuleSet>('/tiers/rules');
}

/** saveTierRules replaces the whole ordered set; a revision other than the stored one is a 409. */
export function saveTierRules(revision: number, rules: TierRuleInput[]): Promise<TierRulesSaved> {
  return api<TierRulesSaved>('/tiers/rules', { method: 'PUT', body: { revision, rules } });
}

export async function getTierPresets(): Promise<TierPreset[]> {
  return (await api<TierPreset[] | null>('/tiers/presets')) ?? [];
}

export async function getTierFields(): Promise<TierField[]> {
  return ((await api<TierField[] | null>('/tiers/fields')) ?? []).map(uniqueSuggestions);
}

/**
 * uniqueSuggestions lists each suggested value once. Several suppliers can offer the same value (a
 * Plex section from a source's setting and from the library index; a user id that two Seerr
 * servers share): one entry keeps the distinct labels, so the editor has one option (and one React
 * key) per value, and every label the value stands for.
 */
export function uniqueSuggestions(field: TierField): TierField {
  const byValue = new Map<string, TierSuggestion>();
  for (const s of field.suggestions ?? []) {
    const k = `${typeof s.value}:${String(s.value)}`;
    const seen = byValue.get(k);
    if (!seen) {
      byValue.set(k, { ...s });
    } else if (!seen.label.split(' / ').includes(s.label)) {
      seen.label = `${seen.label} / ${s.label}`;
    }
  }
  return { ...field, suggestions: [...byValue.values()] };
}

/** previewTiers evaluates the saved rules (rules undefined) or a draft; nothing is written. */
export function previewTiers(body: { rules?: TierRuleInput[]; destinationIds?: number[] }): Promise<TierPreview> {
  return api<TierPreview>('/tiers/preview', { method: 'POST', body });
}

/** listPreviewItems pages a preview's files; an expired preview (10 min) is a 404. */
export async function listPreviewItems(id: string, q: TierPreviewItemQuery): Promise<Paged<TierPreviewItem>> {
  const raw = await api<unknown>(`/tiers/preview/${encodeURIComponent(id)}/items${query({ ...q })}`);
  return toPaged<TierPreviewItem>(raw, q.page, q.pageSize);
}

export async function listFlags(): Promise<ItemFlag[]> {
  return (await api<ItemFlag[] | null>('/tiers/flags')) ?? [];
}

export function addFlag(body: FlagInput): Promise<ItemFlag> {
  return api<ItemFlag>('/tiers/flags', { method: 'POST', body });
}

export function deleteFlag(id: number): Promise<void> {
  return api<void>(`/tiers/flags/${id}`, { method: 'DELETE' });
}

export function getCatalogFile(id: number): Promise<FileDetail> {
  return api<FileDetail>(`/catalog/files/${id}`);
}

// ---- Release (§8.5, §12.1) ----

/** ReleaseParams are the release fields of POST /destinations/{id}/sync. */
export interface ReleaseParams {
  releaseDemoted?: boolean;
  releaseOf?: number;
  releaseRevision?: number;
}

export interface SyncRequest extends ReleaseParams {
  dryRun: boolean;
  allowChanges: boolean;
}

/** startSync queues a sync of a destination with the release fields when given. */
export function startSync(destinationId: number, body: SyncRequest): Promise<Job> {
  return api<Job>(`/destinations/${destinationId}/sync`, { method: 'POST', body });
}

/**
 * startReleasePreview queues the dry run that lists the kept files a release would free. Nothing
 * is changed; its job is where "Apply release" confirms it.
 */
export function startReleasePreview(destinationId: number): Promise<Job> {
  return startSync(destinationId, { dryRun: true, allowChanges: false, releaseDemoted: true });
}

/** tierRevisionOf reads stats.tierRevision of a sync (0 when it recorded none). */
export function tierRevisionOf(job: Pick<Job, 'stats'>): number {
  const v = job.stats?.tierRevision;
  return typeof v === 'number' && Number.isInteger(v) && v > 0 ? v : 0;
}

/** isReleasePreview reports whether a job is a release preview (a dry-run sync with releaseDemoted). */
export function isReleasePreview(job: Pick<Job, 'type' | 'dryRun' | 'params'>): boolean {
  return job.type === 'sync' && job.dryRun && !!job.params?.releaseDemoted;
}

/**
 * releaseParamsOf gives a follow-up real sync of job the release it stands for: a release preview
 * is confirmed by its own id and the revision it evaluated; a real release (whose held changes are
 * applied) passes on the preview and revision it confirmed. Any other job gives none.
 */
export function releaseParamsOf(job: Pick<Job, 'id' | 'type' | 'dryRun' | 'params' | 'stats'>): ReleaseParams {
  if (job.type !== 'sync' || !job.params?.releaseDemoted) {
    return {};
  }
  if (job.dryRun) {
    return { releaseDemoted: true, releaseOf: job.id, releaseRevision: tierRevisionOf(job) };
  }
  return { releaseDemoted: true, releaseOf: job.params.releaseOf, releaseRevision: job.params.releaseRevision };
}

/**
 * applyRelease starts the real release of a finished release preview: it runs what the guard
 * would hold (allowChanges) and releases only that preview's records, and only while the rules
 * are still at the revision it evaluated (else 409).
 */
export function applyRelease(preview: Pick<Job, 'id' | 'type' | 'dryRun' | 'params' | 'stats'>): Promise<Job> {
  const destinationId = preview.params?.destinationId;
  if (!destinationId || !isReleasePreview(preview)) {
    return Promise.reject(new Error('This job is not a release preview.'));
  }
  const revision = tierRevisionOf(preview);
  if (!revision) {
    return Promise.reject(new Error('This preview recorded no rule revision; run the release preview again.'));
  }
  return startSync(destinationId, { dryRun: false, allowChanges: true, ...releaseParamsOf(preview) });
}

File: internal/tiers/tiers.go
// Package tiers is Bunkarr's tier engine (docs/design/phase2-3.md §8): an ordered rule set that
// decides, per live catalog file and per destination, whether the file is copied (full), only
// listed in the destination's manifests (manifest), or neither (skip), and why.
//
// It owns the tables tier_rules and item_flags and the setting tiers.revision. Tiers are
// evaluated, never stored (D5): the facts come from the catalog and the metadata caches
// (internal/mediaindex) at evaluation time.
//
// Safety rules that live here:
//   - D1 (the user's decision): with no rules every file is full, the fallback after the last
//     rule is full, and a file flagged irreplaceable is full at every destination. The spec's
//     manifest-by-default rule set is only a preset, loaded into the editor, never applied.
//   - S14 (unknown never lowers protection): every condition is true, false or unknown; unknown is
//     never true and its negation is unknown. A rule that is unknown and more protective than the
//     rule that matches later decides instead (the decision is unknownPromoted). Stale caches,
//     missing integrations, unmapped or mismatched paths and conflicting claims (D12) are unknown.
//   - S15 is the syncer's (a tier change never removes a backup); this package only decides.
//
// The evaluator is pure (Evaluator: facts in, decision out). LoadFacts builds the facts of a
// source's files; Engine combines the rules, the facts and the flags for the syncer
// (Decisions), the manifest builder and the API (Preview, FactsFor, Fields).
package tiers

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"
)

// Queryer runs read queries: *sql.DB (the read pool) and *sql.Tx satisfy it, so the manifest
// builder can evaluate tiers inside its read transaction. It has the method set of
// mediaindex.Queryer.
type Queryer interface {
	QueryContext(ctx context.Context, query string, args ...any) (*sql.Rows, error)
	QueryRowContext(ctx context.Context, query string, args ...any) *sql.Row
}

// Tier is what a destination does with a file (design §8, D9).
type Tier string

// Tiers, from the most protective.
const (
	// Full: the file is copied.
	Full Tier = "full"
	// Manifest: the file is not copied, but listed in the destination's manifests.
	Manifest Tier = "manifest"
	// Skip: the file is not copied (an *arr file is still listed in its item).
	Skip Tier = "skip"
)

// Valid reports whether t is a known tier.
func (t Tier) Valid() bool { return t == Full || t == Manifest || t == Skip }

// Protection orders tiers: full > manifest > skip (S14).
func (t Tier) Protection() int {
	switch t {
	case Full:
		return 3
	case Manifest:
		return 2
	case Skip:
		return 1
	}
	return 0
}

// Match says how a rule's conditions combine.
type Match string

// Matches.
const (
	// MatchAll: false if any condition is false, true if all are true, otherwise unknown.
	MatchAll Match = "all"
	// MatchAny: true if any condition is true, false if all are false, otherwise unknown.
	MatchAny Match = "any"
)

// Result is a tri-state truth value (S14).
type Result int8

// Results.
const (
	False Result = iota
	True
	Unknown
)

// Not negates r; unknown stays unknown.
func (r Result) Not() Result {
	switch r {
	case True:
		return False
	case False:
		return True
	}
	return Unknown
}

// resultOf converts a known boolean.
func resultOf(b bool) Result {
	if b {
		return True
	}
	return False
}

// String renders r as "true", "false" or "unknown".
func (r Result) String() string {
	switch r {
	case True:
		return "true"
	case False:
		return "false"
	}
	return "unknown"
}

// MarshalJSON renders r as "true", "false" or "unknown".
func (r Result) MarshalJSON() ([]byte, error) { return json.Marshal(r.String()) }

// UnmarshalJSON reads "true", "false" or "unknown".
func (r *Result) UnmarshalJSON(b []byte) error {
	var s string
	if err := json.Unmarshal(b, &s); err != nil {
		return err
	}
	switch s {
	case "true":
		*r = True
	case "false":
		*r = False
	case "unknown":
		*r = Unknown
	default:
		return fmt.Errorf("invalid result %q", s)
	}
	return nil
}

// Condition is one test of a rule: {field, op, value} (design §8.2).
type Condition struct {
	Field string          `json:"field"`
	Op    string          `json:"op"`
	Value json.RawMessage `json:"value,omitempty"`
}

// Rule is a stored tier rule (TierRule of design §13).
type Rule struct {
	ID int64 `json:"id"`
	// Priority is the evaluation order, 1..n.
	Priority   int         `json:"priority"`
	Name       string      `json:"name"`
	Enabled    bool        `json:"enabled"`
	Match      Match       `json:"match"`
	Conditions []Condition `json:"conditions"`
	Action     Tier        `json:"action"`
	// DestinationIDs is "Applies at": nil means every destination; a list means only those (the
	// rule is not evaluated elsewhere); an empty list means none.
	DestinationIDs []int64   `json:"destinationIds"`
	CreatedAt      time.Time `json:"createdAt"`
	UpdatedAt      time.Time `json:"updatedAt"`
}

// AppliesAt reports whether the rule is evaluated at destination destID.
func (r Rule) AppliesAt(destID int64) bool {
	if r.DestinationIDs == nil {
		return true
	}
	for _, id := range r.DestinationIDs {
		if id == destID {
			return true
		}
	}
	return false
}

// RuleInput is a rule as PUT /tiers/rules and the presets carry it. ID 0 is a new rule.
type RuleInput struct {
	ID         int64       `json:"id,omitempty"`
	Name       string      `json:"name"`
	Enabled    *bool       `json:"enabled,omitempty"`
	Match      Match       `json:"match,omitempty"`
	Conditions []Condition `json:"conditions"`
	Action     Tier        `json:"action"`
	// DestinationIDs: null (or absent) is every destination, [] none.
	DestinationIDs []int64 `json:"destinationIds"`
}

// enabled returns the input's enabled flag (default true).
func (in RuleInput) enabled() bool { return in.Enabled == nil || *in.Enabled }

// match returns the input's match (default all).
func (in RuleInput) match() Match {
	if in.Match == "" {
		return MatchAll
	}
	return in.Match
}

// RuleSet is the stored rules in priority order and their revision (GET /tiers/rules).
type RuleSet struct {
	Revision int64  `json:"revision"`
	Rules    []Rule `json:"rules"`
}

// ReasonSource names where a fact came from: kind is catalog, arr, plex, tautulli, seerr,
// maintainerr or flag; IntegrationID the integration, when there is one.
type ReasonSource struct {
	Kind          string `json:"kind"`
	IntegrationID int64  `json:"integrationId,omitempty"`
}

// Reason is one evaluated condition (design §8.1): what was compared (Actual), the result, and,
// for an unknown result, why. A Seerr user appears by id only. RuleID 0 with field
// flag.irreplaceable is the built-in irreplaceable override.
type Reason struct {
	RuleID         int64           `json:"ruleId"`
	ConditionIndex int             `json:"conditionIndex"`
	Field          string          `json:"field"`
	Op             string          `json:"op"`
	Value          json.RawMessage `json:"value,omitempty"`
	Actual         any             `json:"actual,omitempty"`
	Result         Result          `json:"result"`
	Source         ReasonSource    `json:"source"`
	Why            string          `json:"why,omitempty"`
}

// Built-in rule names (rule id 0).
const (
	// FallbackRuleName names the built-in fallback: no rule matched, so the file is full.
	FallbackRuleName = "no rule matched"
	// IrreplaceableRuleName names the built-in override for flagged files.
	IrreplaceableRuleName = "irreplaceable"
)

// Decision is a file's tier at one destination and why (design §8.1).
type Decision struct {
	Tier Tier `json:"tier"`
	// RuleID and RuleName name the deciding rule; 0 is a built-in (the fallback or the
	// irreplaceable override).
	RuleID   int64  `json:"ruleId"`
	RuleName string `json:"ruleName"`
	// Reasons are the deciding rule's conditions that made it decide (for an unknownPromoted
	// decision: its unknown conditions).
	Reasons []Reason `json:"reasons"`
	// Unknown are the unknown conditions of every unknown rule evaluated before the decision.
	Unknown []Reason `json:"unknown"`
	// UnknownPromoted is set when the tier is only this protective because a more protective rule
	// could not be decided (S14): a copy that exists only because of it is a change for the
	// mass-change guard (S10).
	UnknownPromoted bool `json:"unknownPromoted"`
	// Revision is the tiers.revision the rules had.
	Revision int64 `json:"revision"`
	// Follows is the path (inside the source) of the file whose decision this is: the media file
	// of a sidecar, or the most protective name of a hardlink group.
	Follows string `json:"follows,omitempty"`
}

// FallbackDecision is the built-in fallback: full, no rule matched.
func FallbackDecision(revision int64) Decision {
	return Decision{Tier: Full, RuleName: FallbackRuleName, Reasons: []Reason{}, Unknown: []Reason{}, Revision: revision}
}

// UnknownWhy returns the first reason why a fact of the decision is unknown ("Radarr cache is
// 31 h old"), for the hold messages of unknown-promoted copies.
func (d Decision) UnknownWhy() string {
	for _, r := range d.Reasons {
		if r.Result == Unknown && r.Why != "" {
			return r.Why
		}
	}
	for _, r := range d.Unknown {
		if r.Why != "" {
			return r.Why
		}
	}
	return "a fact is unknown"
}

// Summary renders a decision in one line for job logs: `manifest (rule "Keep tagged")`.
func (d Decision) Summary() string {
	var b strings.Builder
	b.WriteString(string(d.Tier))
	switch {
	case d.RuleID == 0 && d.RuleName == IrreplaceableRuleName:
		b.WriteString(" (irreplaceable)")
	case d.RuleID == 0:
		b.WriteString(" (" + FallbackRuleName + ")")
	default:
		fmt.Fprintf(&b, " (rule %q)", d.RuleName)
	}
	if d.UnknownPromoted {
		b.WriteString("; unknown: " + d.UnknownWhy())
	}
	if d.Follows != "" {
		b.WriteString("; follows " + d.Follows)
	}
	return b.String()
}

// Limits of a rule set (design §8.2).
const (
	// MaxRules is the most rules a rule set holds.
	MaxRules = 100
	// MaxConditions is the most conditions a rule holds.
	MaxConditions = 32
	// MaxNameLen is the longest rule name, in characters.
	MaxNameLen = 100
)

// ErrRevisionMismatch means a save was based on another revision of the rules (409).
var ErrRevisionMismatch = errors.New("the tier rules changed since they were loaded; reload them and try again")

// ErrNotFound means a flag (or a preview) does not exist.
var ErrNotFound = errors.New("not found")

// ErrConflict means a flag already exists.
var ErrConflict = errors.New("conflict")

// ValidationError is an invalid rule, condition or flag (400). RuleIndex and ConditionIndex are
// 0-based positions in the request (-1: not about one).
type ValidationError struct {
	RuleIndex      int
	ConditionIndex int
	Msg            string
}

// Error implements error.
func (e *ValidationError) Error() string {
	switch {
	case e.RuleIndex >= 0 && e.ConditionIndex >= 0:
		return fmt.Sprintf("rule %d, condition %d: %s", e.RuleIndex+1, e.ConditionIndex+1, e.Msg)
	case e.RuleIndex >= 0:
		return fmt.Sprintf("rule %d: %s", e.RuleIndex+1, e.Msg)
	}
	return e.Msg
}

func invalid(rule, cond int, format string, args ...any) error {
	return &ValidationError{RuleIndex: rule, ConditionIndex: cond, Msg: fmt.Sprintf(format, args...)}
}

// Warning is a rule value that no fresh index knows (a renamed tag or profile, design §8.7): the
// condition then evaluates false (or unknown) everywhere, silently without the warning.
type Warning struct {
	RuleIndex      int    `json:"ruleIndex"`
	ConditionIndex int    `json:"conditionIndex"`
	Message        string `json:"message"`
}

// UnknownSource is a cache whose facts are unknown now, and why (the preview's banner).
type UnknownSource struct {
	IntegrationID int64  `json:"integrationId"`
	Name          string `json:"name"`
	Reason        string `json:"reason"`
}
